### ISSUE TITLE
recompile with -fPIC linker error for frida-gum ABI x86_64

### ISSUE THREAD

**Issue by thejunkjon** [frida/frida-gum#244]
Hi,

When I use the following build (https://github.com/frida/frida/releases/download/10.6.24/frida-gum-devkit-10.6.24-android-x86_64.tar.xz) and attempt to link libfrida-gum.a against my android application, I get the error 

"libfrida-gum.a(unix64.o): requires dynamic R_X86_64_PC32 reloc against '_frida_ffi_closure_unix64_inner' which may overflow at runtime; recompile with -fPIC)"

All the other ABIs for android (arm, arm64 and x86) work fine.  I am linking all the libraries the same way.  For some reason though, the x86_64 build doesn't link :(

**Comment by thejunkjon** [comment]
Here's the entire message.

FAILED: && /opt/android/sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++  --target=x86_64-none-linux-android --gcc-toolchain=/opt/android/sdk/ndk-bundle/toolchains/x86_64-4.9/prebuilt/linux-x86_64 --sysroot=/opt/android/sdk/ndk-bundle/sysroot -fPIC -isystem /opt/android/sdk/ndk-bundle/sysroot/usr/include/x86_64-linux-android -D__ANDROID_API__=23 -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -Wa,--noexecstack -Wformat -Werror=format-security   -O0 -fno-limit-debug-info  -Wall -Wextra -Weverything -Weffc++ -Wno-c++98-compat-pedantic -Wno-padded -fno-omit-frame-pointer  -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libatomic.a --sysroot /opt/android/sdk/ndk-bundle/platforms/android-23/arch-x86_64 -Wl,--build-id -Wl,--warn-shared-textrel -Wl,--fatal-warnings -Wl,--no-undefined -Wl,-z,noexecstack -Qunused-arguments -Wl,-z,relro -Wl,-z,now -shared -Wl,-soname 

I am using Android NDK r16.

**Comment by oleavr** [comment]
Please use r15c, or rebuild the SDK with r16 and help sort out any issues. (I don't have the bandwidth to do this right now.)

**Comment by thejunkjon** [comment]
No worries @oleavr . Thanks for the project!!

Actually, I think I might have figured it out.  I tried recompiling with NDK r15c and the issue was still there so I don't think it's an NDK issue.  I think the issue is with the way libffi is linked on x86_64.

```
>> readelf --relocs libffi.a

File: libffi.a(win64.o)                                                                                                                                   
                                                                                                                                                          
Relocation section '.rela.text' at offset 0x618 contains 4 entries:                                                                                       
  Offset          Info           Type           Sym. Value    Sym. Name + Addend                                                                          
000000000081  000600000004 R_X86_64_PLT32    0000000000000000 abort - 4                                                                                   
0000000000d9  000600000004 R_X86_64_PLT32    0000000000000000 abort - 4                                                                                   
0000000000f9  000600000004 R_X86_64_PLT32    0000000000000000 abort - 4                                                                                   
**000000000173  000900000002 R_X86_64_PC32     0000000000000000 ffi_closure_win64_inne - 4**  

File: libffi.a(unix64.o)                                                                                                                                  
                                                                                                                                                          
Relocation section '.rela.text' at offset 0x838 contains 3 entries:                                                                                       
  Offset          Info           Type           Sym. Value    Sym. Name + Addend                                                                          
00000000011b  000500000004 R_X86_64_PLT32    0000000000000000 abort - 4                                                                                   
**0000000001f4  000800000002 R_X86_64_PC32     0000000000000000 ffi_closure_unix64_inn - 4**                                                                  
0000000002a6  000500000004 R_X86_64_PLT32    0000000000000000 abort - 4                                                                                   
                                                                                                                                                          
Relocation section '.rela.eh_frame' at offset 0x880 contains 5 entries:                                                                                   
  Offset          Info           Type           Sym. Value    Sym. Name + Addend                                                                          
000000000020  000100000002 R_X86_64_PC32     0000000000000000 .text + 0                                                                                   
000000000048  000100000002 R_X86_64_PC32     0000000000000000 .text + 15e                                                                                 
000000000060  000100000002 R_X86_64_PC32     0000000000000000 .text + 1b0                                                                                 
00000000007a  000100000002 R_X86_64_PC32     0000000000000000 .text + 2aa                                                                                 
000000000090  000100000002 R_X86_64_PC32     0000000000000000 .text + 2fc                                                                                 
                                                                                                                                                          
Relocation section '.rela.debug_info' at offset 0x8f8 contains 9 entries:                                                                                 
  Offset          Info           Type           Sym. Value    Sym. Name + Addend                                                                          
000000000006  00030000000a R_X86_64_32       0000000000000000 .debug_abbrev + 0                                                                           
00000000000c  00040000000a R_X86_64_32       0000000000000000 .debug_line + 0                                                                             
000000000010  000100000001 R_X86_64_64       0000000000000000 .text + 0                                                                                   
000000000018  000100000001 R_X86_64_64       0000000000000000 .text + 330                                                                                 
0000000000d4  000100000001 R_X86_64_64       0000000000000000 .text + 0                                                                                   
0000000000ff  000100000001 R_X86_64_64       0000000000000000 .text + 15e                                                                                 
000000000126  000100000001 R_X86_64_64       0000000000000000 .text + 1b0                                                                                 
000000000154  000100000001 R_X86_64_64       0000000000000000 .text + 2aa                                                                                 
00000000017e  000100000001 R_X86_64_64       0000000000000000 .text + 2fc                                                                                 
                                                                                                                                                          
Relocation section '.rela.debug_aranges' at offset 0x9d0 contains 2 entries:                                                                              
  Offset          Info           Type           Sym. Value    Sym. Name + Addend                                                                          
000000000006  00020000000a R_X86_64_32       0000000000000000 .debug_info + 0                                                                             
000000000010  000100000001 R_X86_64_64       0000000000000000 .text + 0                                                                                   
                                                                                                                                                          
Relocation section '.rela.debug_line' at offset 0xa00 contains 1 entries:                                                                                 
  Offset          Info           Type           Sym. Value    Sym. Name + Addend                                                                          
000000000052  000100000001 R_X86_64_64       0000000000000000 .text + 0            

``` 
For some reason when this library was built, win64.c and unix64.c were built without PIC.

**Comment by oleavr** [comment]
Oh, nice catch! Would you be interested in digging a bit deeper? To only build libffi, do:

`make -f Makefile.sdk.mk FRIDA_HOST=android-x86_64 build/fs-android-x86_64/lib/pkgconfig/libffi.pc`

**Comment by oleavr** [comment]
The CONFIG_SITE environment variable should take care of most of the customizations, and this might be where things go wrong.

**Comment by thejunkjon** [comment]
I tracked this to a bug in libffi.  I'll submit a PR to fix the issue in their library. The line here (https://github.com/libffi/libffi/blob/master/src/x86/unix64.S#L277) should be "call	PLT(C(ffi_closure_unix64_inner))" and the line here (https://github.com/libffi/libffi/blob/master/src/x86/win64.S#L217) should be "call	PLT(C(ffi_closure_win64_inner))"

**Comment by oleavr** [comment]
Awesome catch! Thanks a lot for digging into this!

Feel free to submit one towards our libffi fork as well, and I can rebuild Frida's SDKs with the fix included.

**Comment by oleavr** [comment]
Your fix is currently being released in 10.6.25. Cheers!

**Comment by thejunkjon** [comment]
Oh great :) Thanks @oleavr
